### PR TITLE
Converted mkrelease script to node

### PR DESCRIPTION
**Pull request by jtpetty**

### CODE REVIEW

File: src/mkrelease.sh [hunk 1]
```diff
@@ -1,132 +0,0 @@
-#!/bin/bash
-
-###############################################################################
-#
-#  ./mkrelease.sh [version]
-#        - Creates a release branch and updates necessary files for creating a
-#          release named [version] of the agent
-#
-###############################################################################
```
Comment (damccorm) [inline src/mkrelease.sh]: Maybe we should keep this around for a bit just in case?

File: release/package.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  "version": "1.0.0",
+  "description": "Flow for releasing agent",
+  "main": "mkrelease.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "Microsoft Corporation",
+  "license": "MIT",
+  "dependencies": {
+    "child_process": "^1.0.2",
```
Comment (damccorm) [inline release/package.json]: Nit: Isn't child_process part of built in node? I don't _think_ we need to explicitly add it here

File: release/package.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "agent-release",
+  "version": "1.0.0",
+  "description": "Flow for releasing agent",
+  "main": "mkrelease.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
```
Comment (damccorm) [inline release/package.json]: Nit: I'd probably cut this, but maybe have a release command so you can just say `npm release` rather than explicitly calling `node mkrelease.js` - I think that's more canonical node

File: release/mkrelease.js [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+    }
+    catch(e)
+    {
+        console.log('Error:', e.stack);
+    }
+}
+
+function createIntegrationFiles(newRelease, callback)
+{
+    mkpath(integrationDir, function (err) {
```
Comment (damccorm) [inline release/mkrelease.js]: Rather than going async and adding this `mkpath` module, can we just use node's built in [mkdirSync](https://nodejs.org/api/fs.html#fs_fs_mkdirsync_path_options) with the recursive option? You need node 10 for the recursive option, but that's also current LTS

File: release/mkrelease.js [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+        cp.execSync("rm  -rf " + integrationDir + "/PublishVSTSAgent-*");
+        versionifySync(__dirname + "/../src/Misc/InstallAgentPackage.template.xml",
+            integrationDir + "/InstallAgentPackage.xml",
+            newRelease
+        );
+        var agentVersionPath=newRelease.replace('.', '-');
+        var publishDir = integrationDir + "/PublishVSTSAgent-" + agentVersionPath
+            mkpath(publishDir, function (err) {
+                if (err) throw err;
+                versionifySync(__dirname + "/../src/Misc/PublishVSTSAgent.template.ps1",
```
Comment (damccorm) [inline release/mkrelease.js]: Nit: Use `path.join` to combine directories

Comment (damccorm) [reply release/mkrelease.js]: That's true here and some other places

File: release/mkrelease.js [hunk 6]
```diff
@@ -0,0 +1,306 @@
+const fs = require('fs');
+const request = require('request');
+const cp = require('child_process');
+const mkpath = require('mkpath');
+const naturalSort = require('natural-sort');
+
+var gitHubRequest = request.defaults({
+    headers: {'User-Agent': 'Request'}
+})
+const integrationDir = __dirname + '/../_layout/integrations';
```
Comment (damccorm) [inline release/mkrelease.js]: Style nit: use uppercase for consts - https://github.com/felixge/node-style-guide#use-uppercase-for-constants

File: release/package.json (hunk 3 above)
Comment (damccorm) [reply release/package.json]: Could also have a couple different commands for different arg patterns

File: release/mkrelease.js [hunk 7]
```diff
@@ -0,0 +1,306 @@
+const fs = require('fs');
+const request = require('request');
```
Comment (damccorm) [inline release/mkrelease.js]: You don't need to change this, but in general I think we should probably prefer libraries/patterns that allow for async/await. Notably in this case, we could've used our own stuff - https://github.com/microsoft/typed-rest-client - which is also good for dogfooding. Async/await lets us avoid the callbacks, which can get messy pretty fast (and kinda do in this PR IMO, its a little hard to follow)

Comment (damccorm) [reply release/mkrelease.js]: If you do want to do this, but with minimal changes, it looks like there are some wrappers around `request` that can get the job done - https://www.npmjs.com/package/request#promises--asyncawait

I'm ok as is though.

File: release/mkrelease.js [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+            versionifySync(__dirname + "/../src/Misc/UnpublishVSTSAgent.template.ps1",
+                publishDir + "/UnpublishVSTSAgent-" + agentVersionPath + ".ps1",
+                newRelease
+            );
+            callback();
+        });
+    });
+}
+
+function execInForground(command, directory)
```
Comment (damccorm) [inline release/mkrelease.js]: Spelling nit: `execInForeground`

Comment (damccorm) [reply release/mkrelease.js]: Any reason we need to specify that its happening in the foreground? Really this is just `execSync` with a little directory/opt support.

File: src/mkrelease.sh (hunk 1 above)
Comment (zachariahcox) [reply src/mkrelease.sh]: it's in git if we need it.

File: release/mkrelease.js (hunk 7 above)
Comment (zachariahcox) [reply release/mkrelease.js]: Danny's comment makes sense. Let's schedule moving to the typed rest client for practice.

File: release/mkrelease.js [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+const naturalSort = require('natural-sort');
+
+var gitHubRequest = request.defaults({
+    headers: {'User-Agent': 'Request'}
+})
+const integrationDir = __dirname + '/../_layout/integrations';
+
+const git = 'git';
+
+const release_re = /^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$/;
```
Comment (zachariahcox) [inline release/mkrelease.js]: this should follow the casing convention used elsewhere in the file.

File: release/mkrelease.js (hunk 8 above)
Comment (zachariahcox) [reply release/mkrelease.js]: agreed, this api needs some doc.

Comment (zachariahcox) [reply release/mkrelease.js]: my guess is that it has a "dryrun" option

File: release/mkrelease.js [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+                    commitADOL2Changes(integrationDir + "/AzureDevOps", newRelease);
+                    commitADOConfigChange(integrationDir + "/AzureDevOps.ConfigChange", newRelease);
+                    console.log('done.');
+                });
+            });
+        }
+    );
+}
+
+main(process.argv.slice(2));
```
Comment (zachariahcox) [inline release/mkrelease.js]: missing eof newline

File: release/mkrelease.js (hunk 8 above)
Comment (jtpetty) [reply release/mkrelease.js]: The issue is that execSync by default does not connect stdin, stdout, stderr to the same as the running node process. For some of the git commands, I wanted the output to go to the terminal. For the spawning of the editor to work in the terminal, I needed stdin and stdout hooked up. I also wanted to have the dry run option in one place. I'll add some comments :)

File: release/package.json (hunk 3 above)
Comment (jtpetty) [reply release/package.json]: I see how to make it:
```
npm run release
```
Is that what you mean?

I think we could actually move the dev.sh and. the other supporting scripts into here. We would need to rename the directory though.

File: release/mkrelease.js (hunk 7 above)
Comment (jtpetty) [reply release/mkrelease.js]: The conversion was pretty straight forward and was happy to get rid of the nested callbacks.

File: release/mkrelease.js [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+    {
+        console.log("Invalid version '" + newRelease + "'. Version must be in the form of <major>.<minor>.<patch> where each level is 0-999");
+        process.exit(-1);
+    }
+    var body = await (await httpc.get(GIT_HUB_API_URL_ROOT + "/releases/tags/v" + newRelease)).readBody();
+    body = JSON.parse(body);
+    if (body.message !== "Not Found")
+    {
+        console.log("Version " + newRelease + " is already in use");
+        process.exit(-1)
```
Comment (damccorm) [inline release/mkrelease.js]: Nit: semicolon

File: release/mkrelease.js [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+
+
+async function verifyNewReleaseTagOk(newRelease)
+{
+    if (newRelease === "" || !newRelease.match(VALID_RELEASE_RE) || newRelease.endsWith('.999.999'))
+    {
+        console.log("Invalid version '" + newRelease + "'. Version must be in the form of <major>.<minor>.<patch> where each level is 0-999");
+        process.exit(-1);
+    }
+    var body = await (await httpc.get(GIT_HUB_API_URL_ROOT + "/releases/tags/v" + newRelease)).readBody();
```
[the remark this answers is not in the document]
Comment (damccorm) [reply release/mkrelease.js]: This is true elsewhere
